### PR TITLE
MRG, ENH: Allow channel names > 15 char

### PR DESCRIPTION
**Pull request by larsoner**
MATLAB half of https://github.com/mne-tools/mne-python/pull/8574. It probably makes the most sense to review and merge these two concurrently, since they have changes that are designed to be equivalent.

Ready for review/merge @agramfort

### CODE REVIEW

File: matlab/fiff_write_ch_infos.m [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    for k=1:length(chs)
+        fiff_start_block(fid,FIFF.FIFFB_CH_INFO);
+        fiff_write_string(fid,FIFF.FIFF_CH_DACQ_NAME,chs(k).ch_name);
+        fiff_end_block(fid, FIFF.FIFFB_CH_INFO)
+    end
+end
+
+return;
+
+end
```
Comment (agramfort) [inline matlab/fiff_write_ch_infos.m:45]: missing empty line

**Review by agramfort — APPROVED** [review]
@ftadel @robertoostenveld do you want to have a look as it will end up affecting brainstorm and fieldtrip?

File: matlab/fiff_read_proj.m [hunk 2]
```diff
@@ -45,6 +45,8 @@
 me='MNE:fiff_read_proj';
 
 if nargin ~= 2
+    rename_struct = struct();
+elseif nargin ~= 3
```
Comment (ftadel) [inline matlab/fiff_read_proj.m:49]: This test sequence is weird. With nargin={0,1,4}, it passes.
The error 'Incorrect number of arguments' is never generated when needed.

**Comment by ftadel** [comment]
What acquisition system generates files with channel names > 15 chars?

**Comment by robertoostenveld** [comment]
I think that in brain vision recorder it would be possible to specify long channel names; at least the file format allows for it. [EDF](https://www.edfplus.info/specs/edf.html) allows for channel names up to 16 characters.

**Comment by robertoostenveld** [comment]
is there an example fif file available with extended-length channel names?

File: matlab/test/test_ex_read_raw.m [hunk 3]
```diff
@@ -26,3 +26,27 @@ function test_ex_read_raw_()
...
 assertTrue(all(times < 53.01))
 assertTrue(all(times > 50.99))
+
+% import mne, numpy as np
+% data = np.random.RandomState(0).randn(3, 2)
+% info = mne.create_info(['MEG1', 'MEG2 with a very long name', 'STI 014'], 1000., ['grad', 'grad', 'stim'])
+% mne.io.RawArray(data, info).save('matlab/test/data/test_long_raw.fif')
+want_data = [[1.76405235, 0.40015721]; [0.97873798, 2.2408932]; [1.86755799, -0.97727788]];
+want_times = [0., 0.001]
+want_names = {'MEG1', 'MEG2 with a very long name', 'STI 014'};
```
Comment (larsoner) [inline matlab/test/test_ex_read_raw.m:36]: @robertoostenveld this test file was generated using the MNE-Python code above:

https://github.com/mne-tools/mne-matlab/pull/21/files#diff-b92edbcdf7fb511560a73a838a821de52f65f958b0910e05cafaff6c75808898

**Review by larsoner — COMMENTED** [review]
A couple of real use cases that have channel names that are greater than 15 characters can also be seen in the tests I had to modify for MNE-Python. Some BDF test file had `'EEG LDAMT_01-REF'`:

https://github.com/mne-tools/mne-python/pull/8574/files#diff-8e562c8ca00f3d8bae7b28e973dd6a9965b82623f7d497cc10bb616db94880c7R411

EGI had `['Resp_Temperature', 'Resp_Effort_Chest', 'Resp_Effort_Abdomen']`:

https://github.com/mne-tools/mne-python/pull/8574/files#diff-02dae7f343ff9f6d66fb7b2a0df262e81a6c8bed3b5eb403d8ba2e6c903ba61f

There are probably others, these are just examples that came up in testing.

File: matlab/test/test_ex_read_raw.m (hunk 3 above)
Comment (robertoostenveld) [reply matlab/test/test_ex_read_raw.m:36]: the link in the comment seems to be self-referring to this comment itself. Where would I find the file?

Comment (robertoostenveld) [reply matlab/test/test_ex_read_raw.m:36]: oh, sorry: I found it!

**Comment by robertoostenveld** [comment]
For me it does not work:

```
>> filename = 'test_long_raw.fif';
>> fiff_define_constants;
>> info = fiff_read_meas_info(filename);

Invalid field name: 'MEG2 with a ver'.
Error in setfield (line 35)
    s.(deblank(strField)) = varargin{end};
Error in fiff_read_extended_ch_info (line 29)
            setfield(rename_struct, ch.ch_name, data);
Error in fiff_read_meas_info (line 160)
[chs, rename_struct] = fiff_read_extended_ch_info(chs, meas_info, fid); 
```

the problem appears to be that there are spaces in the channel name.

**Comment by larsoner** [comment]
Makes sense, though I'm not sure why it works for me locally and on CIs. Maybe a MATLAB version thing?

In any case, would you recommend using container.Map?

https://www.mathworks.com/matlabcentral/answers/113258-is-there-any-concept-like-dictionary-or-hash-talbes-in-matlab-like-in-python

My MATLAB is pretty rusty... I could brute force it probably with strcmp and two lists if necessary. Basically I need a way to map abbreviated (15 char) names to full names.

**Comment by ftadel** [comment]
> In any case, would you recommend using container.Map?

I've never used it, but it's standard Matlab after R2008b so no reason for not using it.

I would probably go for the solution with the Nx2 cell-array of strings, but I can see why a Python mind would prefer a dictionary :)

**Comment by larsoner** [comment]
Nx2 cell array implemented, @robertoostenveld this works for me in MATLAB and Octave:
```
>> filename = 'test/data/test_long_raw.fif';
>> info = fiff_read_meas_info(filename);
>> info.ch_names

ans =

  1x3 cell array

    {'MEG1'}    {'MEG2 with a very long name'}    {'STI 014'}
```
Feel free to give it a shot!

**Comment by robertoostenveld** [comment]
yep,  it now also works fine on the FieldTrip side

```
filename = 'test/data/test_long_raw.fif';
info = fiff_read_meas_info(filename);
hdr = ft_read_header(filename); 

want_names = {'MEG1', 'MEG2 with a very long name', 'STI 014'};
assert(isequal(hdr.label(:), want_names(:)));
```

**Comment by larsoner** [comment]
MNE-Python PR is merged so I'll get this in, too -- thanks for testing and reviewing so quickly!

**Comment by ftadel** [comment]
Reported in Brainstorm distribution: https://github.com/brainstorm-tools/brainstorm3/commit/cf76e9feeb605f1a0d4838afe24d7383d8d9c618
Thanks

**Pull request merged**